hubble/src/indexer/fetcher.rs

```diff
@@ -0,0 +1,212 @@
+    postgres::{get_current_height, update_block_status, update_current_height},
+};
+
+trait Chunkeable {
+    type T;
+
+    fn range_chunks(self, chunk_size: usize) -> impl Iterator<Item = Range<Self::T>>;
+}
+
+impl Chunkeable for BlockRange {
+    type T = BlockHeight;
+
+    fn range_chunks(
+        self: BlockRange,
+        chunk_size: usize,
+    ) -> impl Iterator<Item = Range<BlockHeight>> {
+        self.clone().step_by(chunk_size).map(move |block_start| {
+            let block_end = (block_start + chunk_size as u64).min(self.end);
+            block_start..block_end
+        })
+    }
+}
+
+impl<T: FetcherClient> Indexer<T> {
+    pub async fn run_fetcher(&self, fetcher_client: T) -> Result<(), Report> {
+        self.run_to_finalized(&fetcher_client)
+            .instrument(info_span!("run-to-finalized"))
+            .await?;
+        self.run_to_tip(&fetcher_client)
+            .instrument(info_span!("run-to-tip"))
+            .await?;
+        Ok(())
+    }
+
+    async fn run_to_finalized(&self, fetcher_client: &T) -> Result<(), Report> {
+        // run to finalized
+        loop {
+            debug!("fetching last finalized block");
+            match fetcher_client.fetch(BlockReference::LastFinalized).await? {
+                None => {
+                    info!("no finalized block => start 'run to tip'");
+                    return Ok(());
+                }
+                Some(last_finalized) => {
+                    let next_height = self.next_height().await?;
+                    if next_height + self.chunk_size as u64 > last_finalized.height()? {
+                        info!("near finalized height (current: {} finalized: {}) => start 'run to tip'", next_height, last_finalized.height()?);
+                        return Ok(());
+                    }
+
+                    let catch_up_range: BlockRange = next_height..last_finalized.height()?;
+                    info!(
+                        "missing blocks: {}..{}",
+                        catch_up_range.start, catch_up_range.end
+                    );
+
+                    for slice in catch_up_range.range_chunks(self.chunk_size).into_iter() {
+                        info!("handling chunk: {}..{}", slice.start, slice.end);
+
+                        let (tx, mut rx) = mpsc::channel(self.chunk_size);
+                        last_finalized
+                            .fetch_range(tx, slice.clone())
+                            .instrument(info_span!("fetch"))
+                            .await?;
+
+                        let mut expected_next_height = slice.start;
+
+                        while let Some(message) = rx.recv().await {
+                            expected_next_height = self
+                                .handle_fetch_message(message, expected_next_height)
+                                .instrument(info_span!("store"))
+                                .await?
+                        }
```

Comment:
More of a style issue, but it is better to have fetch_range return a Stream vs using a channel to pass messages. That will allow you to handle messages concurrently, as well as make debugging easier (slap a .inspect on the stream).

API would then be:

```
let stream = last_finalized.fetch_range(...).instrument(...);
stream.for_each(self.handle_fetch_message).await;
```

Reply:
done